url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/zmq/StreamEngine.java

```diff
@@ -0,0 +1,623 @@
+        in_event ();
+    }
+    
+    private void unplug () {
+        assert (plugged);
+        plugged = false;
+
+        //  Cancel all fd subscriptions.
+        if (io_enabled) {
+            io_object.rm_fd (handle);
+            io_enabled = false;
+        }
+
+        //  Disconnect from I/O threads poller object.
+        io_object.unplug ();
+
+        //  Disconnect from session object.
+        if (encoder != null)
+            encoder.set_msg_source (null);
+        if (decoder != null)
+            decoder.set_msg_sink (null);
+        session = null;
+    }
+    
+// JDK6    @Override
+    public void terminate () 
+    {
+        if (!terminating && encoder != null && encoder.has_data ())
+        {
+            terminating = true;
+            return;
+        }
+        unplug ();
+        destroy ();
+    }
+
+// JDK6    @Override
+    public void in_event () 
+    {
+        
+        //  If still handshaking, receive and process the greeting message.
+        if (handshaking)
+            if (!handshake ())
+                return;
+        
+        assert (decoder != null);
+        boolean disconnection = false;
+
+        //  If there's no data to process in the buffer...
+        if (insize == 0) {
+
+            //  Retrieve the buffer and read as much data as possible.
+            //  Note that buffer can be arbitrarily large. However, we assume
+            //  the underlying TCP layer has fixed buffer size and thus the
+            //  number of bytes read will be always limited.
+            inbuf = decoder.get_buffer ();
+            insize = read (inbuf);
+            inbuf.flip();
+
+            //  Check whether the peer has closed the connection.
+            if (insize == -1) {
+                insize = 0;
+                disconnection = true;
+            }
+        }
+
+        //  Push the data to the decoder.
+        int processed = decoder.process_buffer (inbuf, insize);
+
+        if (processed == -1) {
+            disconnection = true;
+        }
+        else {
+
+            //  Stop polling for input if we got stuck.
+            if (processed < insize)
+                io_object.reset_pollin (handle);
+
+            //  Adjust the buffer.
+            insize -= processed;
+        }
+
+        //  Flush all messages the decoder may have produced.
+        session.flush ();
+
+        //  An input error has occurred. If the last decoded message
+        //  has already been accepted, we terminate the engine immediately.
+        //  Otherwise, we stop waiting for socket events and postpone
+        //  the termination until after the message is accepted.
+        if (disconnection) {
+            if (decoder.stalled ()) {
+                io_object.rm_fd (handle);
+                io_enabled = false;
+            } else
+                error ();
+        }
+
+    }
+    
+// JDK6    @Override
+    public void out_event () 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 386-402 into a separate method.